table/internal_iterator.h

```diff
@@ -60,12 +60,20 @@ class InternalIteratorBase : public Cleanable {
   // REQUIRES: Valid()
   virtual void Next() = 0;
 
+  // Moves to the next entry in the source, and return result. Iterator
+  // implementation should override this method to help methods inline better,
+  // or when MayBeOutOfUpperBound() is non-trivial.
+  // REQUIRES: Valid()
   virtual bool NextAndGetResult(IterateResult* result) {
     Next();
     bool is_valid = Valid();
     if (is_valid) {
       result->key = key();
-      result->may_be_out_of_upper_bound = MayBeOutOfUpperBound();
+      // Default may_be_out_of_upper_bound to true to avoid unnecessary virtual
+      // call. If an implementation has non-trivial MayBeOutOfUpperBound(),
+      // it should also override NextAndGetResult().
+      result->may_be_out_of_upper_bound = true;
+      assert(MayBeOutOfUpperBound());
```

Comment:
I prefer removing this assert. It's always safe to set `may_be_out_of_upper_bound = true`.

Reply:
I want to avoid people forgetting to override NextAndGetResult() when they should.

Reply:
But even if they forget it, there isn't any correctness issue, right?

Reply:
Anyway, I can merge it and we can continue the conversation. It shouldn't be a blocker because there isn't a new internal iterator yet.

Reply:
If someone ever implement `MayBeOutOfUpperBound()`, it is likely he want its result to be propagated. I think the assert is a must to avoid human error.